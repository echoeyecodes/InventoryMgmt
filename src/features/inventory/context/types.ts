import { Inventory } from "../models/Inventory";

export type AddInventoryParams = Omit<
  Inventory,
  "id" | "createdAt" | "updatedAt"
>;
export type UpdateInventoryParams = {
  id: string;
  data: Omit<Partial<Inventory>, "createdAt" | "updatedAt">;
};

export type InventoryContextType = {
  data: Inventory[];
  addInventory: (inventory: AddInventoryParams) => void;
  updateInventory: (inventory: UpdateInventoryParams) => void;
  removeInventory: (id: string) => void;
  getInventory: (id: string) => Inventory | undefined;
  getInventoryByTitle: (title: string) => Inventory | undefined;
};
