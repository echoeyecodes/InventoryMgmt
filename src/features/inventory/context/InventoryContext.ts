import { createContext } from "react";
import { InventoryContextType } from "./types";

export const InventoryContext = createContext<InventoryContextType>({
  data: [],
  addInventory: () => {},
  updateInventory: () => {},
  removeInventory: () => {},
  getInventory: () => undefined,
  getInventoryByTitle: () => undefined,
});
