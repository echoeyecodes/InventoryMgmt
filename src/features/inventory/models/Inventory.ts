export type Inventory = {
  id: string;
  title: string;
  description: string;
  quantity: number;
  price: number;
  createdAt: string;
  updatedAt: string;
};
