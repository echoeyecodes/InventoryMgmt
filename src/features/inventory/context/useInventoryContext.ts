import { useContext } from "react";
import { InventoryContext } from "./InventoryContext";

export const useInventoryContext = () => {
  return useContext(InventoryContext);
};
