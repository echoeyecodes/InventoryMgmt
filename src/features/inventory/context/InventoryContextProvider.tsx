import {
  PropsWithChildren,
  useEffect,
  useMemo,
  useState,
} from "react";
import { InventoryContext } from "./InventoryContext";
import { Inventory } from "../models/Inventory";
import storage from "@/lib/storage";
import { AddInventoryParams, UpdateInventoryParams } from "./types";
import { ActivityIndicator } from "react-native";

const INVENTORY_KEY = "inventory_db";

export const InventoryContextProvider = (props: PropsWithChildren) => {
  const [isLoading, setIsLoading] = useState(true);
  const [data, setData] = useState<Inventory[]>([]);

  const addInventory = (inventory: AddInventoryParams) => {
    const now = new Date().toISOString();
    const newData: Inventory[] = [
      ...data,
      {
        id: now,
        createdAt: now,
        updatedAt: now,
        ...inventory,
      },
    ];
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const removeInventory = (id: string) => {
    const newData = data.filter((inventory) => inventory.id !== id);
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const updateInventory = (inventory: UpdateInventoryParams) => {
    const now = new Date().toISOString();
    const newData = data.map((item) =>
      item.id === inventory.id
        ? { ...item, ...inventory.data, updatedAt: now }
        : item
    );
    setData(newData);
    storage.setItem(INVENTORY_KEY, JSON.stringify(newData));
  };

  const getInventory = (id: string) => {
    return data.find((inventory) => inventory.id === id);
  };

  const getInventoryByTitle = (title: string) => {
    return data.find(
      (inventory) =>
        inventory.title.trim().toLowerCase() === title.trim().toLowerCase()
    );
  };

  useEffect(() => {
    const fetchData = async () => {
      setIsLoading(true);
      const data = await storage.getItem(INVENTORY_KEY);
      setData(data ? JSON.parse(data) : []);
      setIsLoading(false);
    };
    fetchData();
  }, []);

  const sortedData = useMemo(() => {
    const sortedData = data.sort((a: Inventory, b: Inventory) => {
      return new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime();
    });
    return sortedData;
  }, [data]);

  return (
    <InventoryContext.Provider
      value={{
        data: sortedData,
        getInventoryByTitle,
        addInventory,
        updateInventory,
        removeInventory,
        getInventory,
      }}
    >
      {isLoading ? <ActivityIndicator /> : props.children}
    </InventoryContext.Provider>
  );
};
