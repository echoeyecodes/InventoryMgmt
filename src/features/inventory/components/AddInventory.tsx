import { DefaultLayout } from "@/components/display/DefaultLayout";
import {
  ThemedText,
  ThemedTextProps,
} from "@/components/typography/ThemedText";
import { spacing } from "@/constants/spacing";
import { Alert, StyleSheet, View } from "react-native";
import Feather from "@expo/vector-icons/Feather";
import { Button, IconButton } from "@/components/button";
import { useRouter } from "expo-router";
import { Divider } from "@/components/Divider";
import { TextInput } from "@/components/input/TextInput";
import { ScrollView } from "@/components/list/ScrollView";
import { useTranslation } from "@/lib/i18n";
import { z } from "zod";
import useFormWithValidator from "@/lib/react-form/useForm";
import { Controller } from "react-hook-form";
import { FormWarningLabel } from "@/components/typography/FormWarningLabel";
import { useThemeColor } from "@/lib/hooks/colors/useThemeColor";
import { useInventoryContext } from "../context/useInventoryContext";
import { Inventory } from "../models/Inventory";

const schema = z.object({
  title: z.string().min(1),
  description: z.string().min(3),
  quantity: z.coerce.number().int().positive().min(1),
  price: z.coerce.number().int().positive().min(100),
});

const FormTitle = (props: ThemedTextProps) => {
  return <ThemedText variant="h6" fontSize={14} {...props} />;
};

export const AddInventory = ({ data }: { data?: Inventory }) => {
  const {
    addInventory,
    updateInventory,
    removeInventory,
    getInventoryByTitle,
  } = useInventoryContext();
  const router = useRouter();
  const form = useFormWithValidator<z.infer<typeof schema>>({
    schema,
    defaultValues: {
      title: data?.title ?? "",
      description: data?.description ?? "",
      quantity: data?.quantity ?? 1,
      price: data?.price ?? 1,
    },
  });
  const colors = useThemeColor();
  const { t } = useTranslation();

  const onSubmit = (newData: z.infer<typeof schema>) => {
    if (!isEditMode) {
      const isNameUnique = !getInventoryByTitle(newData.title);
      if (!isNameUnique) {
        return form.setError("title", {
          message: t("inventory.inventory_exists"),
        });
      }
      addInventory({
        title: newData.title,
        description: newData.description,
        quantity: newData.quantity,
        price: newData.price,
      });
    } else {
      if (data) {
        updateInventory({
          id: data.id,
          data: {
            title: newData.title,
            description: newData.description,
            quantity: newData.quantity,
            price: newData.price,
          },
        });
      }
    }
    router.back();
  };

  const handleDeleteInventory = () => {
    if (data) {
      Alert.alert(t("common.delete"), t("inventory.delete_confirm"), [
        {
          text: t("common.cancel"),
          style: "cancel",
        },
        {
          text: t("common.delete"),
          onPress: () => {
            removeInventory(data.id);
            router.back();
          },
        },
      ]);
    }
  };

  const isEditMode = !!data;

  return (
    <DefaultLayout style={styles.container}>
      <View style={styles.header}>
        <IconButton onPress={router.back}>
          <Feather name="chevron-left" size={24} color={colors.text} />
        </IconButton>
        <ThemedText variant="h5" testID="add-inventory-title">
          {isEditMode
            ? t("inventory.update_inventory")
            : t("inventory.add_inventory")}
        </ThemedText>
      </View>

      <Divider />

      <ScrollView
        style={styles.form}
        contentContainerStyle={styles.formContent}
      >
        <View style={styles.formItem}>
          <FormTitle>{t("inventory.inventory_input_title")}</FormTitle>
          <Controller
            control={form.control}
            name="title"
            render={({ field: { onChange, onBlur, value } }) => (
              <TextInput
                placeholder={t("inventory.inventory_input_title_placeholder")}
                onChangeText={onChange}
                testID="inventory-input-title"
                onBlur={onBlur}
                value={value}
              />
            )}
          />
          {form.formState.errors.title && (
            <FormWarningLabel>
              {form.formState.errors.title.message}
            </FormWarningLabel>
          )}
        </View>

        <View style={styles.formItem}>
          <FormTitle>{t("inventory.inventory_input_description")}</FormTitle>
          <Controller
            control={form.control}
            name="description"
            render={({ field: { onChange, onBlur, value } }) => (
              <TextInput
                placeholder={t(
                  "inventory.inventory_input_description_placeholder"
                )}
                onChangeText={onChange}
                onBlur={onBlur}
                testID="inventory-input-description"
                value={value}
              />
            )}
          />
          {form.formState.errors.description && (
            <FormWarningLabel>
              {form.formState.errors.description.message}
            </FormWarningLabel>
          )}
        </View>

        <View style={styles.formItem}>
          <FormTitle>{t("inventory.inventory_input_quantity")}</FormTitle>
          <Controller
            control={form.control}
            name="quantity"
            render={({ field: { onChange, onBlur, value } }) => (
              <TextInput
                placeholder={t(
                  "inventory.inventory_input_quantity_placeholder"
                )}
                onChangeText={onChange}
                testID="inventory-input-quantity"
                onBlur={onBlur}
                keyboardType="numeric"
                value={value.toString()}
              />
            )}
          />
          {form.formState.errors.quantity && (
            <FormWarningLabel>
              {form.formState.errors.quantity.message}
            </FormWarningLabel>
          )}
        </View>

        <View style={styles.formItem}>
          <FormTitle>{t("inventory.inventory_input_price")}</FormTitle>
          <Controller
            control={form.control}
            name="price"
            render={({ field: { onChange, onBlur, value } }) => (
              <TextInput
                placeholder={t("inventory.inventory_input_price_placeholder")}
                onChangeText={onChange}
                onBlur={onBlur}
                keyboardType="numeric"
                value={value.toString()}
                testID="inventory-input-price"
              />
            )}
          />
          {form.formState.errors.price && (
            <FormWarningLabel>
              {form.formState.errors.price.message}
            </FormWarningLabel>
          )}
        </View>

        <Button
          testID="inventory-submit-button"
          onPress={form.handleSubmit(onSubmit)}
        >
          {isEditMode ? t("common.update") : t("common.save")}
        </Button>

        {isEditMode && (
          <Button
            testID="inventory-delete-button"
            variant="text"
            onPress={handleDeleteInventory}
          >
            {t("common.delete")}
          </Button>
        )}
      </ScrollView>
    </DefaultLayout>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  header: {
    padding: spacing.md,
    flexDirection: "row",
    alignItems: "center",
    gap: spacing.xs,
  },
  form: {
    flex: 1,
  },
  formContent: {
    gap: spacing.md,
    paddingVertical: spacing.md,
    paddingHorizontal: spacing.md,
  },
  formItem: {
    gap: spacing.xs,
  },
});
