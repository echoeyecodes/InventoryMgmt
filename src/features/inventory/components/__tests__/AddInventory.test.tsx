import React from "react";
import { fireEvent, render, waitFor } from "@testing-library/react-native";
import { InventoryContextProvider } from "../../context/InventoryContextProvider";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { NavigationContainer } from "@react-navigation/native";
import { AddInventory } from "../AddInventory";
import { Alert } from "react-native";
import { Inventory } from "../../models/Inventory";

jest.mock("@react-native-async-storage/async-storage", () => ({
  setItem: jest.fn(),
  getItem: jest.fn(),
  removeItem: jest.fn(),
  clear: jest.fn(),
}));

const mockedSetItem = AsyncStorage.setItem as jest.MockedFunction<
  typeof AsyncStorage.setItem
>;

const mockGoBack = jest.fn();

const mockPush = jest.fn();

jest.mock("expo-router", () => ({
  useLocalSearchParams: () => ({
    id: null,
  }),
  useRouter: () => ({
    push: mockPush,
    back: mockGoBack,
  }),
}));

jest.mock("react-native", () => {
  const reactNative = jest.requireActual("react-native");
  reactNative.Alert.alert = jest.fn();
  return reactNative;
});

describe("<AddInventory />", () => {
  it("saves the inventory", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));
    const { getByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const titleInput = getByTestId("inventory-input-title");
      const descriptionInput = getByTestId("inventory-input-description");
      const quantityInput = getByTestId("inventory-input-quantity");
      const priceInput = getByTestId("inventory-input-price");

      const submitButton = getByTestId("inventory-submit-button");

      fireEvent.changeText(titleInput, "Item 1");
      fireEvent.changeText(descriptionInput, "Test description");
      fireEvent.changeText(quantityInput, "5");
      fireEvent.changeText(priceInput, "1000");
      fireEvent.press(submitButton);

      expect(AsyncStorage.setItem).toHaveBeenCalled();
      expect(mockGoBack).toHaveBeenCalled();
    });
  });

  it("renders correctly in add mode", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { getByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const titleInput = getByTestId("inventory-input-title");
      const descriptionInput = getByTestId("inventory-input-description");
      const quantityInput = getByTestId("inventory-input-quantity");
      const priceInput = getByTestId("inventory-input-price");
      const submitButton = getByTestId("inventory-submit-button");

      expect(titleInput).toBeTruthy();
      expect(descriptionInput).toBeTruthy();
      expect(quantityInput).toBeTruthy();
      expect(priceInput).toBeTruthy();
      expect(submitButton).toBeTruthy();
    });
  });

  it("renders correctly in edit mode", async () => {
    const data: Inventory = {
      id: "1",
      title: "Item 1",
      description: "Test description",
      quantity: 5,
      price: 1000,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify([data])
    );

    const { getByTestId, getByText } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory data={data} />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const titleInput = getByTestId("inventory-input-title");
      const descriptionInput = getByTestId("inventory-input-description");
      const quantityInput = getByTestId("inventory-input-quantity");
      const priceInput = getByTestId("inventory-input-price");

      const deleteButton = getByTestId("inventory-delete-button");

      expect(titleInput).toBeTruthy();
      expect(titleInput.props.value).toBe("Item 1");
      expect(descriptionInput).toBeTruthy();
      expect(descriptionInput.props.value).toBe("Test description");
      expect(quantityInput).toBeTruthy();
      expect(quantityInput.props.value).toBe("5");
      expect(priceInput).toBeTruthy();
      expect(priceInput.props.value).toBe("1000");
      expect(getByText("Update Inventory")).toBeTruthy();
      expect(deleteButton).toBeTruthy();
    });
  });

  it("updates the inventory", async () => {
    const inventory: Inventory = {
      id: "112",
      title: "Item 1",
      description: "Test description",
      quantity: 5,
      price: 1000,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify([inventory])
    );
    const { getByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory data={inventory} />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const quantityInput = getByTestId("inventory-input-quantity");
      const priceInput = getByTestId("inventory-input-price");

      const submitButton = getByTestId("inventory-submit-button");

      expect(quantityInput).toBeTruthy();
      expect(priceInput).toBeTruthy();

      fireEvent.changeText(quantityInput, "10");
      fireEvent.changeText(priceInput, "2000");
      fireEvent.press(submitButton);

      const [_, value] = mockedSetItem.mock.calls[3];

      const storedData = JSON.parse(value);

      expect(storedData).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            quantity: 10,
            price: 2000,
          }),
        ])
      );
      expect(mockGoBack).toHaveBeenCalled();
    });
  });

  it("shows the delete alert dialog when delete button is pressed", async () => {
    const data: Inventory = {
      id: "1",
      title: "Item 1",
      description: "Test description",
      quantity: 5,
      price: 1000,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify([data])
    );

    const { getByTestId } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory data={data} />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      const deleteButton = getByTestId("inventory-delete-button");
      fireEvent.press(deleteButton);

      expect(Alert.alert).toHaveBeenCalled();
    });
  });

  it("matches snapshot in add mode", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(JSON.stringify([]));

    const { toJSON } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });

  it("matches snapshot in edit mode", async () => {
    (AsyncStorage.getItem as jest.Mock).mockResolvedValue(
      JSON.stringify([
        {
          id: "1",
          title: "Item 1",
          description: "Test description",
          quantity: 5,
          price: 1000,
        },
      ])
    );

    const { toJSON } = render(
      <NavigationContainer>
        <InventoryContextProvider>
          <AddInventory />
        </InventoryContextProvider>
      </NavigationContainer>
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
