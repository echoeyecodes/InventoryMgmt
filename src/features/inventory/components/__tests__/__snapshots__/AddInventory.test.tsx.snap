// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddInventory /> matches snapshot in add mode 1`] = `<ActivityIndicator />`;

exports[`<AddInventory /> matches snapshot in edit mode 1`] = `<ActivityIndicator />`;
